live/client.py

```diff
@@ -132,3 +134,58 @@ def stop_event(self, event_id):
         # Send request and do exception handling
         self.send_request(http_method="POST", url=url,
                           headers=headers, body=body)
+
+    def find_devices_in_use(self):
+        events_url = f'{self.server_ip}/live_events'
+        events_headers = self.generate_headers(events_url)
+        events_xml = self.send_request(
+            http_method="GET", url=events_url, headers=events_headers)
+
+        events_list = ET.fromstring(events_xml.text)
+
+        # Find in use devices from all events
+        in_use_devices = set()
+        for event in events_list:
+            status = event.find('status').text
+            for input in event.findall('input'):
+                device_input = input.find('device_input')
+                if device_input:
+                    device_name = device_input.find('device_name')
+                    if status in ('preprocessing', 'running'):
+                        in_use_devices.add(device_name.text)
+
+        return in_use_devices
+
+    def get_input_devices(self):
+        devices_url = f'{self.server_ip}/devices'
+        devices_headers = self.generate_headers(devices_url)
+        devices_xml = self.send_request(
+            http_method="GET", url=devices_url, headers=devices_headers)
+
+        devices_list = ET.fromstring(devices_xml.text)
+
+        # Find all devices info
+        all_devices = set()
+        for device in devices_list:
+            all_devices.add(device.find('device_name').text)
+
+        devices_info_dict = xmltodict.parse(devices_xml.text)[
+            'device_list']['device']
+
+        in_use_devices = self.find_devices_in_use()
+
+        devices_availability = {}
+
+        for d in all_devices:
+            devices_availability[d] = True
```

Comment:
You could also use a dictionary comprehension to  generate the `available_devices`. Something like:

```python
>>> all_devices = ["dev1", "dev2", "dev3", "dev4", "dev5"]
>>> devices_in_use = ["dev1", "dev3"]
>>> {d:(d not in devices_in_use) for d in all_devices}
{'dev3': False, 'dev2': True, 'dev1': False, 'dev5': True, 'dev4': True}
```

Reply:
cool!